src/rulesets-new/lifecycle-rules.ts

```diff
@@ -0,0 +1,112 @@
+import { RuleError, Ruleset, SpecificationRule } from "@useoptic/rulesets-base";
+import { links } from "../docs";
+import { stabilityKey } from "./constants";
+
+const stabilityRequirement = new SpecificationRule({
+  name: "resource stability",
+  docsLink: links.versioning.stabilityLevels,
+  matches: (specification, rulesContext) =>
+    rulesContext.specification.change !== "removed",
+  rule: (specificationAssertions) => {
+    specificationAssertions.requirement(
+      "be provided for every resource document",
+      (specification) => {
+        const stability: string | undefined = specification.value[stabilityKey];
+        const allowed = ["wip", "experimental", "beta", "ga"];
+        if (!stability || !allowed.includes(stability)) {
+          throw new RuleError({
+            message: `${stability} must be one of allowed values ${allowed.join(
+              ", ",
+            )}`,
+          });
+        }
+      },
+    );
+  },
+});
+
+const stabilityTransitions = new SpecificationRule({
+  name: "resource stability transitions",
+  docsLink: links.versioning.promotingStability,
+  rule: (specificationAssertions) => {
+    specificationAssertions.changed(
+      "not change unless it was wip",
+      (before, after) => {
+        const beforeStability: string | undefined = before.value[stabilityKey];
```

Comment:
What is the shape of `before` here?

How would I know to look in `before.value` for the top-level extension? Looking at the type, it seems like it's an annotated `any` type (unless I read something wrong).

Thinking about this from these perspectives:
- How I would write a specification rule that required the top-level OpenAPI 3 property `tags:` to be a non-empty array?
- How would I make an assertion on the URLs in the `servers:` section?

Reply:
before + after are the same shape (i.e. they are both specification facts) - on a changed value, we have both before and after, other assertions only have a single value.

Each value has the following shape (depending on where they are):
```
specificationAssertions."assertion"('condition', (specification)) 

specification = {
  location: - location data for the "fact"
  value: - fact value
  raw: - the raw value at the the fact (i.e. the raw OpenAPI as JSON at that point)
}
```

> How I would write a specification rule that required the top-level OpenAPI 3 property tags: to be a non-empty array?
```javascript
specificationAssertions.requirement(specification => mustBeArray(specification.value.tags))
```

> How would I make an assertion on the URLs in the servers: section?
```javascript
specificationAssertions.requirement(specification => someAssertion(specification.value.servers))
```